public/app/features/alerting/unified/utils/rule-id.ts

```diff
@@ -299,6 +299,34 @@ export function hashQuery(query: string) {
   // labels matchers can be reordered, so sort the enitre string, esentially comparing just the character counts
   return query.split('').sort().join('');
 }
+export function hashQuery(query: string): string {
+  // Step 1: Remove enclosing parentheses if present
+  if (query.length > 1 && query[0] === '(' && query[query.length - 1] === ')') {
+    query = query.slice(1, -1);
+  }
+
+  // Step 2: Remove whitespace and newlines
+  query = query.replace(/\s|\n/g, '');
+
+  // Step 3: Count character occurrences
+  const charCounts = new Map<number, number>(); // Use a map to support UTF-16 code units
+  for (let i = 0; i < query.length; i++) {
```

Comment:
According to MDN it's probably better to simply use the `for...of` statement so we don't have to worry about surrogate pairs :)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt#looping_with_codepointat

This simplifies the code here a bit:

```
// Step 3: Count character occurrences
  const frequency = new Map<string, number>();
  for (const char of query) {
    const code = char.codePointAt(0)!;
    frequency.set(char, (frequency.get(char) ?? 0) + 1);
  }
```